src/commons.rs

```diff
@@ -246,53 +166,125 @@ pub extern "C" fn z_qos_default() -> z_qos_t {
-    let Some(sample) = sample else {
-        return zc_payload_null();
-    };
-    let buf = unsafe { std::mem::transmute::<_, &ZBuf>(sample._zc_buf).clone() };
-    zc_owned_payload_t {
-        payload: sample.payload,
-        _owner: unsafe { std::mem::transmute(buf) },
+pub extern "C" fn z_sample_keyexpr(sample: &z_sample_t) -> z_keyexpr_t {
+    (&sample.key_expr).into()
+}
+/// The encoding of the payload.
+#[no_mangle]
+pub extern "C" fn z_sample_encoding(sample: &z_sample_t) -> z_encoding_t {
+    (&sample.encoding).into()
+}
+/// The sample's data, the return value aliases the sample.
+///
+/// If you need ownership of the buffer, you may use `z_sample_owned_payload`.
+#[no_mangle]
+pub extern "C" fn z_sample_payload(sample: &z_sample_t) -> z_bytes_t {
+    sample.payload.slices().next().unwrap_or(b"").into()
+}
+/// Returns the sample's payload after incrementing its internal reference count.
+///
+/// Note that other samples may have received the same buffer, meaning that mutating this buffer may
+/// affect the samples received by other subscribers.
+#[no_mangle]
+pub extern "C" fn z_sample_owned_payload(sample: &z_sample_t) -> z_owned_buffer_t {
+    sample.payload.clone().into()
+}
+/// The sample's kind (put or delete).
+#[no_mangle]
+pub extern "C" fn z_sample_kind(sample: &z_sample_t) -> z_sample_kind_t {
+    sample.kind.into()
+}
+/// The samples timestamp
+#[no_mangle]
+pub extern "C" fn z_sample_timestamp(sample: &z_sample_t) -> z_timestamp_t {
+    sample.timestamp.as_ref().into()
+}
+/// The qos with which the sample was received.
+#[no_mangle]
+pub extern "C" fn z_sample_qos(sample: &z_sample_t) -> z_qos_t {
+    sample.qos.into()
+}
+/// The sample's attachment.
+///
+/// `sample` is aliased by the return value.
+#[no_mangle]
+pub extern "C" fn z_sample_attachment(sample: &z_sample_t) -> z_attachment_t {
+    match &sample.attachment {
+        Some(attachment) => z_attachment_t {
+            data: attachment as *const _ as *mut c_void,
+            iteration_driver: Some(attachment_iteration_driver),
+        },
+        None => z_attachment_null(),
     }
 }
 
+/// An owned sample.
+///
+/// This is a read only type that can only be constructed by cloning a `z_sample_t`.
+/// Like all owned types, its memory must be freed by passing a mutable reference to it to `zc_sample_drop`.
+#[repr(C)]
+pub struct zc_owned_sample_t {
+    _0: z_owned_keyexpr_t,
+    _1: z_owned_buffer_t,
+    _2: z_owned_buffer_t,
+    _3: [usize; 12],
+}
+
+impl_guarded_transmute!(Option<Sample>, zc_owned_sample_t);
+
+/// Clone a sample in the cheapest way available.
+#[no_mangle]
+pub extern "C" fn zc_sample_clone(sample: &z_sample_t) -> zc_owned_sample_t {
+    Some(sample.deref().clone()).into()
+}
+
+/// Returns `true` if `sample` is valid.
+///
+/// Note that there exist no fallinle constructors for `zc_owned_sample_t`, so validity is always guaranteed
```

Comment:
For completeness I think we need `z_sample_null` - zenoh-cpp will need it at least, for move constructors